analyzer/runtime/evm.go

```diff
@@ -290,3 +292,50 @@ func EVMDownloadTokenBalance(ctx context.Context, logger *log.Logger, source sto
 		return nil, fmt.Errorf("download stale token balance type %v not handled", tokenType)
 	}
 }
+
+// EVMMaybeUnmarshalEncryptedData breaks down a possibly encrypted data +
+// result into their encryption envelope fields. If the data is not encrypted,
+// it returns nils.
+func EVMMaybeUnmarshalEncryptedData(data []byte, result *[]byte) (*int, *[]byte, *[]byte, *[]byte, *[]byte, *[]byte, error) {
+	var call sdkTypes.Call
+	var publicKey []byte
+	var dataNonce []byte
+	var dataData []byte
+	var callResult sdkTypes.CallResult
+	var resultNonce []byte
+	var resultData []byte
+	if cbor.Unmarshal(data, &call) != nil {
+		// Invalid CBOR means it's bare Ethereum format data. This is normal.
+		// https://github.com/oasisprotocol/oasis-sdk/blob/runtime-sdk/v0.3.0/runtime-sdk/modules/evm/src/lib.rs#L626
+		return nil, nil, nil, nil, nil, nil, nil
+	}
+	if result != nil {
+		if err := cbor.Unmarshal(*result, &callResult); err != nil {
+			return nil, nil, nil, nil, nil, nil, fmt.Errorf("unmarshal outer call result: %w", err)
+		}
+	}
+	switch call.Format {
+	case sdkTypes.CallFormatEncryptedX25519DeoxysII:
+		var callEnvelope sdkTypes.CallEnvelopeX25519DeoxysII
+		if err := cbor.Unmarshal(call.Body, &callEnvelope); err != nil {
+			return nil, nil, nil, nil, nil, nil, fmt.Errorf("outer call format %s unmarshal body: %w", call.Format, err)
+		}
+		publicKey = callEnvelope.Pk[:]
+		dataNonce = callEnvelope.Nonce[:]
```

Comment:
Here and below: Why do we need to copy? Actually, unlike in Python, this [does not copy](https://go.dev/play/p/9fpRLK-i-xd); if the syntax is intentional, what does it achieve?

Reply:
the public key and nonce are arrays, a kind of fixed-size value in Go. the envelope is specific to a certain encryption scheme, where these sizes are known, e.g. a 15-byte nonce for deoxys-ii. this makes a slice that we can put in the db along with potentially different schemes, in case we later come up with a new "callformat"

Reply:
> Here and below: Why do we need to copy? Actually, unlike in Python, this [does not copy](https://go.dev/play/p/9fpRLK-i-xd); if the syntax is intentional, what does it achieve?

You should update `a` to be an array in your example (currently it is a slice). Then it might make more sense:
```golang
a := [3]int{1, 2, 3}
b1 := a    // creates a new array that is a copy of a
b2 := a[:] // creates a slice that references a
```

Reply:
Thank you both! I didn't realize the underlying data was array-typed.